README.rst
.. These are examples of badges you might want to add to your README:
   please update the URLs accordingly

    .. image:: https://api.cirrus-ci.com/github/<USER>/kindle-highlight-exporter.svg?branch=main
        :alt: Built Status
        :target: https://cirrus-ci.com/github/<USER>/kindle-highlight-exporter
    .. image:: https://readthedocs.org/projects/kindle-highlight-exporter/badge/?version=latest
        :alt: ReadTheDocs
        :target: https://kindle-highlight-exporter.readthedocs.io/en/stable/
    .. image:: https://img.shields.io/coveralls/github/<USER>/kindle-highlight-exporter/main.svg
        :alt: Coveralls
        :target: https://coveralls.io/r/<USER>/kindle-highlight-exporter
    .. image:: https://img.shields.io/pypi/v/kindle-highlight-exporter.svg
        :alt: PyPI-Server
        :target: https://pypi.org/project/kindle-highlight-exporter/
    .. image:: https://img.shields.io/conda/vn/conda-forge/kindle-highlight-exporter.svg
        :alt: Conda-Forge
        :target: https://anaconda.org/conda-forge/kindle-highlight-exporter
    .. image:: https://pepy.tech/badge/kindle-highlight-exporter/month
        :alt: Monthly Downloads
        :target: https://pepy.tech/project/kindle-highlight-exporter
    .. image:: https://img.shields.io/twitter/url/http/shields.io.svg?style=social&label=Twitter
        :alt: Twitter
        :target: https://twitter.com/kindle-highlight-exporter

.. image:: https://img.shields.io/badge/-PyScaffold-005CA0?logo=pyscaffold
    :alt: Project generated with PyScaffold
    :target: https://pyscaffold.org/

=========================
kindle-highlight-exporter
=========================


    Have you ever wanted to save the text you've highlighted on your kindle to a neat text file? This repo is for you!


* I wanted to be able to export highlights I made while reading to a neat google doc or comparable.
* Amazon has something that `lets you save highlights from certain books <https://www.howtogeek.com/449256/how-to-backup-your-kindle-highlights-and-notes/>`_, but not DRM-free books. (Almost none of the books on my kindle are DRM protected)
* There's already a `tiny web app which accomplishes the same thing <https://erkanerol.github.io/kindle-clippings-extractor/>`_, but the UI and resulting text output are not as neat as I would like.


.. _pyscaffold-notes:

Note
====

This project has been set up using PyScaffold 4.2.1. For details and usage
information on PyScaffold see https://pyscaffold.org/.


CHANGELOG.rst
=========
Changelog
=========

Version 0.1
===========

- Init repo

tests/conftest.py
"""
    Dummy conftest.py for kindle_highlight_exporter.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    - https://docs.pytest.org/en/stable/fixture.html
    - https://docs.pytest.org/en/stable/writing_plugins.html
"""

# import pytest


tests/test_skeleton.py
import pytest

from kindle_highlight_exporter.skeleton import fib, main

__author__ = "boscacci"
__copyright__ = "boscacci"
__license__ = "MIT"


def test_fib():
    """API Tests"""
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)


def test_main(capsys):
    """CLI Tests"""
    # capsys is a pytest fixture that allows asserts agains stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    main(["7"])
    captured = capsys.readouterr()
    assert "The 7-th Fibonacci number is 13" in captured.out


tests/test_kindledump.py
import pytest

from kindle_highlight_exporter.kindledump import main
from kindle_highlight_exporter.kindledump import import_kindle_textfile
from kindle_highlight_exporter.kindledump import get_clippings_by_author
from kindle_highlight_exporter.kindledump import get_clippings_by_title
from kindle_highlight_exporter.subroutines import fuzzy_matches

__author__ = "boscacci"
__copyright__ = "boscacci"
__license__ = "MIT"

file_loc = "tests/test_data/2021-06-07_My Clippings.txt"
clippings_list = import_kindle_textfile(file_loc)


def test_import_kindle_textfile():
    """API Tests"""
    assert isinstance(clippings_list, list)
    assert len(clippings_list) > 0


def test_fuzzy_matches():
    assert (
        fuzzy_matches(
            value_from_clip=clippings_list[0]["author"],
            valid_values_list=["carreyrou john"],
        )
        == True
    )


def test_get_clippings_by_title():
    clippings_that_matter = get_clippings_by_title(
        clippings_list,
        only_these=[
            "something that matters",
        ],
    )
    assert len(clippings_that_matter) == 174

    hemingway_clips = get_clippings_by_title(
        clippings_list, only_these=["a farewell to arms"]
    )

    assert len(hemingway_clips) == 3


def test_get_clippings_by_author():
    carreyrou_clippings = get_clippings_by_author(
        clippings_list,
        only_these=[
            "John Carreyrou",
        ],
    )
    assert len(carreyrou_clippings) == 4


def test_get_clippings_by_not_author():
    not_carreyrou_clippings = get_clippings_by_author(
        clippings_list,
        exclude_these=[
            "John Carreyrou",
        ],
    )
    assert len(not_carreyrou_clippings) < len(clippings_list)
    assert "being large" in not_carreyrou_clippings[-2]["highlight_text"]


def test_search_by_author_cli(capsys):
    """CLI Tests"""
    # capsys is a pytest fixture that allows asserts agains stdout/stderr
    # https://docs.pytest.org/en/stable/capture.html
    main(
        [
            file_loc,
            "--author",
            "John Carreyrou",
        ]
    )
    captured = capsys.readouterr()
    assert (
        "firewall between the Journal’s editorial and newsroom staffs"
        in captured.out
    )


def test_search_by_title_cli(capsys):
    main(
        [
            file_loc,
            "--title",
            "something that matters",
        ]
    )
    captured = capsys.readouterr()
    assert (
        "everyone reading this book has the potential to make a difference"
        in captured.out
    )


src/kindle_highlight_exporter/skeleton.py
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following lines in the
``[options.entry_points]`` section in ``setup.cfg``::

    console_scripts =
         fibonacci = kindle_highlight_exporter.skeleton:run

Then run ``pip install .`` (or ``pip install -e .`` for editable mode)
which will install the command ``fibonacci`` inside your current environment.

Besides console scripts, the header (i.e. until ``_logger``...) of this file can
also be used as template for Python modules.

Note:
    This skeleton file can be safely removed if not needed!

References:
    - https://setuptools.pypa.io/en/latest/userguide/entry_point.html
    - https://pip.pypa.io/en/stable/reference/pip_install
"""

import argparse
import logging
import sys

from kindle_highlight_exporter import __version__

__author__ = "boscacci"
__copyright__ = "boscacci"
__license__ = "MIT"

_logger = logging.getLogger(__name__)


# ---- Python API ----
# The functions defined in this section can be imported by users in their
# Python scripts/interactive interpreter, e.g. via
# `from kindle_highlight_exporter.skeleton import fib`,
# when using this Python module as a library.


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for _i in range(n - 1):
        a, b = b, a + b
    return a


# ---- CLI ----
# The functions defined in this section are wrappers around the main Python
# API allowing them to be called directly from the terminal as a CLI
# executable/script.


def parse_args(args):
    """Parse command line parameters

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--help"]``).

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(description="Just a Fibonacci demonstration")
    parser.add_argument(
        "--version",
        action="version",
        version="kindle-highlight-exporter {ver}".format(ver=__version__),
    )
    parser.add_argument(dest="n", help="n-th Fibonacci number", type=int, metavar="INT")
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(
        level=loglevel, stream=sys.stdout, format=logformat, datefmt="%Y-%m-%d %H:%M:%S"
    )


def main(args):
    """Wrapper allowing :func:`fib` to be called with string arguments in a CLI fashion

    Instead of returning the value from :func:`fib`, it prints the result to the
    ``stdout`` in a nicely formatted message.

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--verbose", "42"]``).
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m kindle_highlight_exporter.skeleton 42
    #
    run()


src/kindle_highlight_exporter/subroutines.py
from thefuzz import fuzz


def get_loc_values(loc_str):
    if "-" in loc_str:
        start = loc_str.split("-")[0]
        end = loc_str.split("-")[1]
        return int(start), int(end)
    else:
        return int(loc_str), int(loc_str)


def parse_one_clipping(clip):
    """
    Attempt to parse out relevant data for one clipping

    Parameters
    ----------
    clip : str
        Mess of metadata about book/article/etc.

    Returns
    -------
    dict
        Data dict of relevant info from clipping
    """
    clip_dict = dict()
    lines = clip.split("\n")
    clip_dict["source_title"] = lines[0].rsplit("(", 1)[0].strip()

    # Author
    try:
        clip_dict["author"] = lines[0].rsplit("(")[-1].rsplit(")")[0]
        # This indicates that something went wrong getting author:
        if clip_dict["author"] == clip_dict["source_title"]:
            clip_dict["author"] = None
    except IndexError:
        # Title and author are not obvious/separate
        clip_dict["author"] = None

    # !TODO clip_dict['datetime'] = # SOMETHING

    # Page / Location
    try:
        location_str = lines[1].split("Location ")[1].split(" |")[0]
        clip_dict["loc_start"], clip_dict["loc_end"] = get_loc_values(
            location_str
        )
    except IndexError:
        # This uses pages and not loc?
        try:
            clip_dict["page_num"] = lines[1].split("page ")[1].split(" |")[0]
        except IndexError:
            breakpoint()

    clip_dict["highlight_text"] = lines[3]
    return clip_dict


def import_kindle_textfile(file_loc):
    """
    Loads a text file from kindle "my clippings"

    Parameters
    ----------
    file_loc : str
        What is the filepath/filename of my clippings?

    Returns
    -------
    list[dict]
        List of dictionaries representing individual clippings
    """
    with open(file_loc, "r") as infile:
        all_clippings = infile.read()
    split_by_ufeff = all_clippings.split("\ufeff")
    clippings = [clip for clip in split_by_ufeff if clip]
    unsplit_last_one = clippings.pop()
    clippings.extend(unsplit_last_one.split("\n==========\n"))
    clippings = [clip for clip in clippings if clip]
    clip_dicts = [parse_one_clipping(clip) for clip in clippings]
    return clip_dicts


def get_clippings_by_author(
    clip_dicts,
    only_these=None,
    exclude_these=None,
):
    return get_clippings_by_field(
        clip_dicts,
        field="author",
        only_these=only_these,
        exclude_these=exclude_these,
    )


def get_clippings_by_title(
    clip_dicts,
    only_these=None,
    exclude_these=None,
):
    return get_clippings_by_field(
        clip_dicts,
        field="source_title",
        only_these=only_these,
        exclude_these=exclude_these,
    )


def get_clippings_by_field(
    clip_dicts,
    field,
    only_these=None,
    exclude_these=None,
):
    """
    Given a filter field, find relevant clips

    Parameters
    ----------
    clip_dicts : list
        clip dicts to search thru
    only_these : str or list[str], optional
        Which things we want clips for, by default None
    exclude_these : str or list[str], optional
        Which things to exclude from results, by default None

    Returns
    -------
    list[dict]
        list of clippings that match criteria
    """
    # Inclusive approach
    if only_these:
        clip_dicts = [
            clip
            for clip in clip_dicts
            if fuzzy_matches(clip[field], only_these)
        ]

    # Exclusive approach
    if exclude_these:
        clip_dicts = [
            clip for clip in clip_dicts if clip[field] not in exclude_these
        ]
    return clip_dicts


def fuzzy_matches(value_from_clip, valid_values_list):
    scores = [
        fuzz.token_sort_ratio(value_from_clip, valid)
        for valid in valid_values_list
    ]
    if any([score > 75 for score in scores]):
        return True
    return False


src/kindle_highlight_exporter/kindledump.py
"""
To run this script uncomment the following lines in the
``[options.entry_points]`` section in ``setup.cfg``::

    console_scripts =
         kindledump = kindle_highlight_exporter.kindledump:run

Then run ``pip install .`` (or ``pip install -e .`` for editable mode)
which will install the command ``fibonacci`` inside your current environment.

Besides console scripts, the header (i.e. until ``_logger``...) of this file can
also be used as template for Python modules.

Note:
    This skeleton file can be safely removed if not needed!

References:
    - https://setuptools.pypa.io/en/latest/userguide/entry_point.html
    - https://pip.pypa.io/en/stable/reference/pip_install
"""

import argparse
import logging
import sys
import json

from kindle_highlight_exporter import __version__

__author__ = "boscacci"
__copyright__ = "boscacci"
__license__ = "MIT"

_logger = logging.getLogger(__name__)

# ---- Python API ----
# The functions defined in this section can be imported by users in their
# Python scripts/interactive interpreter, e.g. via
# `from kindle_highlight_exporter.kindledump import import_kindle_textfile`,
# when using this Python module as a library.

from kindle_highlight_exporter.subroutines import import_kindle_textfile
from kindle_highlight_exporter.subroutines import get_clippings_by_author
from kindle_highlight_exporter.subroutines import get_clippings_by_title

# ---- CLI ----
# The functions defined in this section are wrappers around the main Python
# API allowing them to be called directly from the terminal as a CLI
# executable/script.


def parse_args(args):
    """Parse command line parameters

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--help"]``).

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Dump kindle data in a nice format"
    )
    parser.add_argument(
        "--version",
        action="version",
        version="kindle-highlight-exporter {ver}".format(ver=__version__),
    )
    parser.add_argument(
        dest="clippings_path",
        help="path of clippings file",
        type=str,
        metavar="CLIPPINGS_PATH",
    )
    parser.add_argument(
        "--author",
        dest="authors",
        help="author names",
        nargs="+",
        type=str,
        metavar="AUTHORS",
    )
    parser.add_argument(
        "--title",
        dest="titles",
        help="titles of books",
        type=str,
        nargs="+",
        metavar="TITLES",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO,
    )
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG,
    )
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(
        level=loglevel,
        stream=sys.stdout,
        format=logformat,
        datefmt="%Y-%m-%d %H:%M:%S",
    )


def main(args):
    """Wrapper allowing :func:`kindledump` to be called with string arguments in a CLI fashion

    Instead of returning the value from :func:`kindledump`, it prints the result to the
    ``stdout`` in a nicely formatted message, or saves to a file.

    Args:
      args (List[str]): command line parameters as list of strings
          (for example  ``["--author", "John Carreyrou"]``).
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting kindle dump")
    args_dict = vars(args)
    clip_dicts = import_kindle_textfile(args_dict["clippings_path"])

    if args_dict.get("authors") and args_dict.get("titles"):
        # ! TODO
        print("Haven't coded this capability yet.")

    elif args_dict.get("authors"):
        authors = args_dict["authors"]
        author_clips = get_clippings_by_author(
            clip_dicts=clip_dicts,
            only_these=authors,
        )
        authors_output_json = json.dumps(author_clips)
        print(authors_output_json)

    elif args_dict.get("titles"):
        titles = args_dict["titles"]
        title_clips = get_clippings_by_title(
            clip_dicts=clip_dicts,
            only_these=titles,
        )
        with open("tests/output_test.json", "w") as jfile:
            json.dump({"data": title_clips}, jfile)

    else:
        print(
            "\nYou didn't enter any search items (with e.g. --author "
            "or --title). Here's what's available:\n"
        )
        authors_titles = {(c["author"], c["source_title"]) for c in clip_dicts}
        author_title_dicts = [
            {"author": a_t[0], "title": a_t[1]} for a_t in authors_titles
        ]
        for a_t_dict in author_title_dicts:
            print(
                f"Author: {a_t_dict['author']}\n"
                f"Title: {a_t_dict['title']}\n\n"
            )

    _logger.info("Script ends here")


def run():
    """Calls :func:`main` passing the CLI arguments extracted from :obj:`sys.argv`

    This function can be used as entry point to create console scripts with setuptools.
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    # ^  This is a guard statement that will prevent the following code from
    #    being executed in the case someone imports this file instead of
    #    executing it as a script.
    #    https://docs.python.org/3/library/__main__.html

    # After installing your project with pip, users can also run your Python
    # modules as scripts via the ``-m`` flag, as defined in PEP 338::
    #
    #     python -m kindle_highlight_exporter.skeleton 42
    #
    run()
